packages/diagram/lib/diagram.dart

```diff
@@ -0,0 +1,442 @@
+    _pixelRatio = pixelRatio;
+    handleMetricsChanged();
+  }
+
+  /// Determines the dimensions of the virtual screen that the diagram will
+  /// be drawn on, in logical units.
+  Size get screenDimensions => _screenDimensions;
+  Size _screenDimensions = _kDefaultDiagramViewportSize;
+  set screenDimensions(Size screenDimensions) {
+    _screenDimensions = screenDimensions;
+    handleMetricsChanged();
+  }
+
+  /// Returns the singleton for the binding, after making sure it is
+  /// initialized.
+  static WidgetsBinding instance() {
+    if (WidgetsBinding.instance == null) {
+      new _DiagramFlutterBinding();
+    }
+    return WidgetsBinding.instance;
+  }
+
+  @override
+  ViewConfiguration createViewConfiguration() {
+    return new _DiagramViewConfiguration(
+      pixelRatio: pixelRatio,
+      size: screenDimensions,
+    );
+  }
+
+  /// Captures an image of the [RepaintBoundary] with the given key.
+  Future<ui.Image> takeSnapshot() async {
+    RenderRepaintBoundary object = _boundaryKey.currentContext.findRenderObject();
+    return await object.toImage(pixelRatio: pixelRatio);
+  }
+
+  /// Updates the current diagram with the given builder as the child of the
+  /// root widget, and generates a new frame.
+  void updateDiagram(
+    WidgetBuilder builder, {
+    Duration duration: Duration.zero,
+  }) {
+    Widget rootWidget = new _Diagram(
+      boundaryKey: _boundaryKey,
+      child: new Builder(builder: builder),
+    );
+    attachRootWidget(rootWidget);
+    pump();
+  }
+
+  /// Advances time by the given duration, and generates a new frame.
+  ///
+  /// The [duration] must not be null, or less than [Duration.zero].
+  void pump({Duration duration: Duration.zero}) {
+    assert(duration != null);
+    assert(duration >= Duration.zero);
+    _timestamp += duration;
+
+    handleBeginFrame(_timestamp);
+    handleDrawFrame();
+  }
+}
+
+/// A generator used by [DiagramController] for generating filenames for frames
+/// in animated diagram sequences.
+typedef File AnimationFrameFilenameGenerator(Duration timestamp, int index);
+
+/// A controller for creating diagrams generated by using Flutter widgets.
+///
+/// This is used to configure and create individual image diagrams, as well as
+/// animations of diagrams with control over the time used by animation tickers.
+class DiagramController {
+  /// Creates a diagram controller.
+  ///
+  /// The [builder] parameter must not be null, and is required.
+  DiagramController({
+    @required WidgetBuilder builder,
+    this.outputDirectory,
+    AnimationFrameFilenameGenerator frameFilenameGenerator,
+    double pixelRatio: 1.0,
+    Size screenDimensions: _kDefaultDiagramViewportSize,
+  }) : assert(builder != null) {
+    this.frameFilenameGenerator = frameFilenameGenerator;
+    outputDirectory ??= Directory.current;
+    _binding.pixelRatio = pixelRatio;
+    _binding.screenDimensions = screenDimensions;
+    this.builder = builder;
+  }
```

Comment:
The idiomatic thing would be to construct the binding here and assert that it's the right type

Reply:
see e.g. https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/binding.dart#L702
(but with an additional assert)